### PR TITLE
feat: support configuring default task images [DET-4079]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
It is not possible to configure the default task images, pull policy and registry credentials
via the `master.yaml`. Additionally made it configurable via the Helm chart as well.

## Test Plan
- [x] Add unit tests
- [x] Test on a local cluster
- [x] Test on a kubernetes cluster

### CODE REVIEW

**Comment by rb-determined-ai** [comment]
At a high level, why are we adding individual default fields rather than just full-blown default templates (which we had planned on doing a long time ago... I'm not sure what actually happened to it though)

**Comment by aaron276h** [comment]
@rb-determined-ai I do agree long term we probably want to make it possible to set defaults for all experiment configurations, however I didn't want to take that work on in the scope of this ticket given that this is a time-sensitive customer request.

**Review by seanr15 — APPROVED** [review]

File: docs/reference/cluster-config.txt [hunk 1]
```diff
@@ -288,6 +288,32 @@ The master supports the following configuration settings:
...
+    vs. CPU agents differently by specifying a dict with two keys, ``cpu`` and ``gpu``.
+    Default values:
+
+    - ``determinedai/environments:py-3.6.9-pytorch-1.4-tf-1.15-cpu-0.5.0`` for CPU agents
+    - ``determinedai/environments:cuda-10.0-pytorch-1.4-tf-1.15-gpu-0.5.0`` for GPU agents.
+
+  - ``force_pull_image``: Defines the default policy for forcibly pulling images from
+    the docker registry and bypassing the docker cache. If a pull policy is specified
+    in the :ref:`experiment config <exp-environment-image>` this default value is
+    overriden. Defaults to ``false``.
```
Comment (armandmcqueen) [inline docs/reference/cluster-config.txt:304]: As of November 1st, unauthenticated users will be capped at 100 pulls per 6 hours. This limit is based on IP address. We might want to call out that `force_pull_image=True` could lead to that cap being hit very quickly.

https://www.docker.com/blog/scaling-docker-to-serve-millions-more-developers-network-egress/

File: master/pkg/model/experiment_config_test.go [hunk 2]
```diff
@@ -415,3 +417,122 @@ func TestExperiment(t *testing.T) {
 	assert.NilError(t, json.Unmarshal(json2, &config3))
 	assert.DeepEqual(t, config1, config3)
 }
+
+func TestMasterConfigImage(t *testing.T) {
```
Comment (rb-determined-ai) [inline master/pkg/model/experiment_config_test.go:421]: Thank you for thorough unit tests!

File: master/pkg/model/defaults.go [hunk 3]
```diff
@@ -23,13 +23,27 @@ const (
...
 		petname.Generate(TaskNameGeneratorWords, TaskNameGeneratorSep))
 
-	return ExperimentConfig{
+	cpuImage := defaultCPUImage
+	gpuImage := defaultGPUImage
+
+	if taskContainerDefaults != nil && taskContainerDefaults.Image != nil {
+		cpuImage = taskContainerDefaults.Image.CPU
+		gpuImage = taskContainerDefaults.Image.GPU
+	}
```
Comment (rb-determined-ai) [inline master/pkg/model/defaults.go:44]: non-blocking: why not just put this at the bottom with the other `if taskContainerDefaults != nil` clause?  Or move that clause up here and take a similar strategy?

File: docs/reference/helm-config.txt [hunk 4]
```diff
@@ -136,12 +136,27 @@ Helm, the deployment should be configured by editing the ``values.yaml`` and
...
     See :ref:`custom-pod-specs` for details.
 
+  - ``cpuImage``: Sets the default docker image for all non-gpu tasks. If a docker image
+    is specified in the :ref:`experiment config <exp-environment-image>`
+    this default is overriden. Default to:
+    ``determinedai/environments:py-3.6.9-pytorch-1.4-tf-1.15-cpu-0.5.0``.
+
+  - ``gpuImage``: Sets the default docker image for all gpu tasks. If a docker image
+    is specified in the :ref:`experiment config <exp-environment-image>`
+    this default is overriden. Default to:
```
Comment (rb-determined-ai) [inline docs/reference/helm-config.txt:157]: Maybe "Default to" -> "Defaults to"?

"Default to" sounds like a command to the reader.

(also above)

**Review by rb-determined-ai — COMMENTED** [review]
I think this looks "good" given the timeline we have to ship this feature.  But adding parameters to a "DefaultExperimentConfig()" function is an awful code smell, since it's not really "default" anymore, is it?

Have you tested how this interacts with templates?

**Review by aaron276h — COMMENTED** [review]
@rb-determined-ai 

>  But adding parameters to a "DefaultExperimentConfig()" function is an awful code smell, since it's not really "default" anymore, is it?

I this it's fine this way it this way since it not returns the default experiment config for this particular cluster. The one time it's not the case is when called from the `Scan()` which is reading an already created config from the database already. If you want we can create a separate function called `applyClusterDefaults` which applies the the `taskContainerDefaults` to the default config. 

> Have you tested how this interacts with templates?

Yes and it works as expected

File: master/pkg/model/defaults.go [hunk 5]
```diff
@@ -97,13 +103,27 @@ func DefaultExperimentConfig() ExperimentConfig {
...
+	if taskContainerDefaults == nil {
+		return defaultConfig
+	}
+
+	defaultConfig.Environment.RegistryAuth = taskContainerDefaults.RegistryAuth
+	defaultConfig.Environment.ForcePullImage = taskContainerDefaults.ForcePullImage
+
+	if taskContainerDefaults.Image != nil {
+		defaultConfig.Environment.Image.CPU = taskContainerDefaults.Image.CPU
+		defaultConfig.Environment.Image.GPU = taskContainerDefaults.Image.GPU
```
Comment (rb-determined-ai) [inline master/pkg/model/defaults.go:125]: Can't you just do `defaultConfig.Environment.Image = taskContainerDefaults.Image`?

File: docs/reference/helm-config.txt [hunk 6]
```diff
@@ -152,12 +152,27 @@ Helm Chart </helm/determined-helm-chart.tgz>`.
       use during distributed training. A valid port range is in the
       format of ``MIN:MAX``.
 
+   -  ``forcePullImage``: Defines the default policy for forcibly pulling images from
+      the docker registry and bypassing the docker cache. If a pull policy is specified
+      in the :ref:`experiment config <exp-environment-image>` this default  is overriden.
+      Defaults to ``false``.
```
Comment (rb-determined-ai) [inline docs/reference/helm-config.txt:158]: update this to include the detail about docker pulls as well?

**Review by rb-determined-ai — APPROVED** [review]

File: docs/reference/helm-config.txt (hunk 6 above)
Comment (aaron276h) [reply docs/reference/helm-config.txt:161]: done

File: master/pkg/model/defaults.go (hunk 5 above)
Comment (aaron276h) [reply master/pkg/model/defaults.go]: good idea

**Pull request merged**